### ISSUE TITLE
Multithreading support

### ISSUE THREAD

**Issue by AlexAkulov** [pierrec/lz4#55]
I noticed that this library uses only one cpu core. It may be problem on processing large data. Do you have ideas about how to add multi thread support?

**Comment by pierrec** [comment]
Probably need to add new types, one for writing (compress) and one for reading (decompress), that handle blocks in parallel. Dont have any time free atm to tackle this, but if you do, feel free to send PRs!

**Comment by pierrec** [comment]
@klauspost thank you for your insights, it looks like an elegant way to deal with concurrent compression
I will probably experiment with your technique and see if this helps for lz4.

**Comment by pierrec** [comment]
First version of the concurrent writer added to branches master and v3.
No release yet as things may move a bit.
If you have time, please try it out and let me know.

**Comment by klauspost** [comment]
@pierrec `10gb.tar` and `rawstudio-mint14.tar` crashes:
```
panic: runtime error: slice bounds out of range [:4194316] with capacity 4194304

goroutine 2450 [running]:
github.com/pierrec/lz4.compressBlock(0xc000d04000, 0x400000, 0x800000, 0xc015e68000, 0x400000, 0x400000, 0xc0067e3f08, 0x10000, 0x10000, 0x0)
        e:/gopath/src/github.com/pierrec/lz4/block.go:164 +0x847
github.com/pierrec/lz4.(*Writer).compressBlock.func2(0xc000d04000, 0x400000, 0x800000, 0xc015e68000, 0x400000, 0x400000, 0xc00666a420, 0xc015a68000, 0x400000, 0x800000, ...)
        e:/gopath/src/github.com/pierrec/lz4/writer.go:289 +0x2ce
created by github.com/pierrec/lz4.(*Writer).compressBlock
        e:/gopath/src/github.com/pierrec/lz4/writer.go:282 +0x49d
```
http://mattmahoney.net/dc/10gb.html and https://files.klauspost.com/compress/rawstudio-mint14.7z

Speed wise it is looking fine compared to S2:

Using test files here: https://github.com/klauspost/compress/tree/master/s2#performance

JSON: Worse compression, but slightly faster than fastest mode.
github-ranks-backup.bin (serialized binary data): Same speed as S2 and better compression.
consensus.db (Binary DB): Slower than S2 and worse compression.
enwik9 (XML): Perfectly in between the 2 S2 modes both in terms of compression and speed.
2GB random data: Same speed/compression

**Comment by klauspost** [comment]
@pierrec Looks nice. 

I noticed is that `Flush()` doesn't seem to keep [its promise](https://github.com/pierrec/lz4/blob/v3/writer.go#L311) any more. AFAICT it only queues the current block.

[`Reset`](https://github.com/pierrec/lz4/blob/v3/writer.go#L377) should probably also clear `z.err`.

**Comment by pierrec** [comment]
@klauspost thanks a lot for this analysis!
I have addressed the issues you mentioned. Running into strange issues when running the tests with concurrency enabled, still investigating.

**Comment by klauspost** [comment]
Looks very good! The crashes are fixed.

`10gb.tar` - Slightly slower than S2, slightly better compression 🆗 
`rawstudio-mint14.tar` - Slightly slower than S2, worse compression. 👎 

So it is definitely "competitive". The margins to S2 are pretty slim in most cases and I think the only difference really comes down to the encoding chosen for each format.

I included LZ4 as a very competitive format for Go compression in a [talk I just gave](https://go-talks.appspot.com/github.com/klauspost/talks/2019/compression.slide#1).
